### PR TITLE
Refactor rendering

### LINKED ISSUE #662: ChunkBatcher named wrong
Based on the thread names, the `ChunkBatcher` are the *threads* created by the class currently named `ChunkBatcher`:

```java
private static final ThreadFactory THREAD_FACTORY = new ThreadFactoryBuilder().setNameFormat("Chunk Batcher %d").setDaemon(true).setUncaughtExceptionHandler(new UncaughtExceptionLogger(LOGGER)).build();
```

### LINKED ISSUE #774: SystemUtil -> ?
`SystemUtil` contains a bunch of static methods. Some of them relate to a "system":

* fine-grained timers
* current operating system
* JVM flags

But many others are completely unrelated:

* patching miscellaneous holes in the Java standard library (`ifPresentOrElse`, `thenCombine`, `completeExceptionally`, `consume`, `get`, `toMap`, ...)
* a bit about the server worker
* reading and writing UUIDs from `Dynamic`s
* an identity hash strategy
* something about translation keys? lol

It appears to be less of a system-related class and more just a dumping ground for random functions that are too small to break into their own named helper class but too broadly applicable to keep near one call site.

Should this class be renamed?

Some suggestions being thrown around on Discord:

* `Misc`
* `Util`
* `Useful`
* `Varia`
* `Helper`
* `Etc`
* ~~Funny comedy suggesions:~~
  * ~~`Junk`~~
  * ~~`Stuff`~~
  * ~~`MiscellaneousGameSystemUtilities`~~
  * ~~`BucketOShi`~~

### LINKED ISSUE #926: Improve concurrency class naming
We currently have a `Mailbox` class and a `MailboxProcessor` class, but a log message in `MailboxProcessor` suggests that in fact `MailboxProcessor` is the `Mailbox`:

```java
    private void schedule() {
        if (hasMessages() && lock()) {
            try {
                executor.execute(this);
            } catch (RejectedExecutionException e1) {
                try {
                    executor.execute(this);
                } catch (RejectedExecutionException e2) {
                    LOGGER.error("Cound not schedule mailbox", e2);
                }
            }
        }
    }
```

I suggest renaming the class currently named `Mailbox` to something like `MessageQueue`, since it's really just a queue which receives messages (for example an action + a priority) and gives an action (for example the highest priority one) when polled. Then, we can rename `MailboxProcessor` (which uses a `MessageQueue` to store its messages, and provides a few extra methods for executing tasks sync/async, locking and unlocking, etc.) to `Mailbox`.

Also, as suggested by the log message in `Actor`, a better name for `createAndSendFutureActor` would be `ask`:

```java
    default <Source> CompletableFuture<Source> ask(Function<? super Actor<Source>, ? extends Msg> messageProvider) {
        CompletableFuture<Source> future = new CompletableFuture<>();
        Actor<Source> actor = create("ask future procesor handle", future::complete); // actor that acts on Source messages to complete the future
        Msg message = messageProvider.apply(actor);
        send(message);
        return future;
    }
```

It creates a future, creates an actor that receives source messages and completes the future, applies a function to convert the actor to a message, and sends that message. So it's really a helper method for sending "questions" to the `Actor` and returning a future for the reply.

The same log message would suggest `Actor` is actually named `ProcessorHandle`, but I think `Actor` is more descriptive (it accepts messages and acts on them), and [already used](https://en.wikipedia.org/wiki/Actor_model) (in fact, [this page](https://berb.github.io/diploma-thesis/original/054_actors.html) matches almost perfectly how the concurrency system works).

### LINKED ISSUE #927: ChunkRenderer -> RenderedChunk
That name would make it clearer that it's storing the actual rendered chunk (the buffers with the vertex data for each layer) rather than just a class that's called to render. It also works better with the `rebuild` method name (it's rebuilding the rendered chunk, not the chunk renderer).

This would also free up the `ChunkRenderer` name for the class currently named `ChunkBatcher` (see https://github.com/FabricMC/yarn/issues/662).

### PR DESCRIPTION
**Pull request by Runemoro**
- `ChunkRenderer` -> `RenderedChunk` (fixes https://github.com/FabricMC/yarn/issues/927)
 - `Mailbox` -> `MessageQueue`, `MailboxProcessor` -> `Mailbox` (fixes https://github.com/FabricMC/yarn/issues/926)
 - `SystemUtil` -> `Util` (fixes https://github.com/FabricMC/yarn/issues/774)
 - `ChunkBatcher` -> `ChunkRenderer` (fixes https://github.com/FabricMC/yarn/issues/662)

### CODE REVIEW

**Comment by liach** [comment]
Uh, this seems to conflict with #917 on rendering stuff

**Comment by Runemoro** [comment]
The only conflict is `LayeredBufferStorage` (I renamed it, and you added methods to it). I'll fix the conflict after your PR is merged.

**Comment by liach** [comment]
Given #929, we might need to hold off on this.

**Comment by Runemoro** [comment]
I'll be reverting the BufferBuilder rename later today so that the other changes can be merged, but leave the issue open for discussion.

**Comment by liach** [comment]
Please update to 19w44a.

**Comment by Runemoro** [comment]
Updated to 19w44a, but don't review yet, I'll revert #929 first.

File: mappings/net/minecraft/util/MessageQueue.mapping
Comment (liach) [inline mappings/net/minecraft/util/MessageQueue.mapping]: This class probably needs a rename as well.

File: mappings/net/minecraft/util/Util.mapping
Comment (liach) [inline mappings/net/minecraft/util/Util.mapping]: If you rename this one to initialize, I think the supplier one below can be renamed as well.

Comment (Runemoro) [reply mappings/net/minecraft/util/Util.mapping]: It would be weird calling the other one `initialize` since it just creates something using a supplier, while this one initializes a given object by applying a consumer to it.

I do agree that they should have the same name since they're used for the same purpose, so I'll name them both `create`.

Comment (liach) [reply mappings/net/minecraft/util/Util.mapping]: Both function as `setup` or so. sure

File: mappings/net/minecraft/client/render/LayeredVertexOutput.mapping
Comment (liach) [inline mappings/net/minecraft/client/render/LayeredVertexOutput.mapping]: Uh no

File: mappings/net/minecraft/client/render/LayeredVertexOutput.mapping
Comment (liach) [inline mappings/net/minecraft/client/render/LayeredVertexOutput.mapping]: No either

File: mappings/net/minecraft/client/render/block/BlockRenderManager.mapping
Comment (liach) [inline mappings/net/minecraft/client/render/block/BlockRenderManager.mapping]: This one is different in that it uses different render layers than those of regular blocks. Beware!

File: mappings/net/minecraft/client/render/chunk/ChunkRenderer.mapping
Comment (liach) [inline mappings/net/minecraft/client/render/chunk/ChunkRenderer.mapping]: What does "important" mean?

Comment (Runemoro) [reply mappings/net/minecraft/client/render/chunk/ChunkRenderer.mapping]: The rebuild task has higher priority, and it will rebuild on the client thread if the option to rebuild near on client thread is enabled. It's used for blocks placed by a player, for example (it's important for those to show up immediately).

File: mappings/net/minecraft/client/render/LayeredVertexOutput.mapping
Comment (Runemoro) [reply mappings/net/minecraft/client/render/LayeredVertexOutput.mapping]: Oops, renamed it at its usage and didn't realize it was a Mojang name. I'll fix this.

Comment (liach) [reply mappings/net/minecraft/client/render/LayeredVertexOutput.mapping]: Well, if mojang did not use this as a method reference at where they render the breaking progress, we could have named it with much more ease instead

File: mappings/net/minecraft/util/MessageQueue.mapping
Comment (liach) [inline mappings/net/minecraft/util/MessageQueue.mapping]:
```suggestion
	CLASS class_3849 Simple
```

File: mappings/net/minecraft/util/MessageQueue.mapping
Comment (liach) [inline mappings/net/minecraft/util/MessageQueue.mapping]:
```suggestion
	CLASS class_3848 Prioritized
```

**Review by liach — APPROVED** [review]

**Comment by liach** [comment]
On a side note, there is a few new additions on rendering details as well.

**Comment by Player3324** [comment]
The Mailbox related changes are ok.

Util is a very generic/common name, maybe it can be aligned more closely with mc.

For ChunkRenderer and ChunkBatcher I'd prefer a stricter interpretation of the term 'rendering' along the lines of drawing, rasterization or getting something to the display/frame buffer. This excludes the cases where the output is some buffer that doesn't get drawn (almost) immediately and discarded, but reused for many draws. Updating or baking chunk(section)s is one way to express this process in the client/render engine context.

File: mappings/net/minecraft/client/render/BuiltChunkStorage.mapping
Comment (Player3324) [inline mappings/net/minecraft/client/render/BuiltChunkStorage.mapping:3]: I think xSize xOffset etc fit our "natural order" naming style better, sizeX etc is prefix style

File: mappings/net/minecraft/util/thread/TaskExecutor.mapping
Comment (Player3324) [inline mappings/net/minecraft/util/thread/TaskExecutor.mapping]: are you sure? isPaused maybe?

File: mappings/net/minecraft/util/thread/ThreadExecutor.mapping
Comment (Player3324) [inline mappings/net/minecraft/util/thread/ThreadExecutor.mapping]: maybe runTasks? execute(Runnable) is already being used above. The same applies to executeTask()

**Review by Player3324 — COMMENTED** [review]
Some mentions of mailbox and message seem to have slipped through, there may also be mentions of "render" in a "building/baking" context.

Overall a very nice change!

File: mappings/net/minecraft/client/render/BuiltChunkStorage.mapping
Comment (Runemoro) [reply mappings/net/minecraft/client/render/BuiltChunkStorage.mapping:3]: Right now we're naming coordinates the other way (see [conventions](https://github.com/FabricMC/yarn/blob/19w44a/CONVENTIONS.md#coordinates), and also almost all current coordinate fields are named this way). The idea is that the field is the "size['s] x [component]".

The other order could work too, but I think it would be better to make a separate issue to change the conventions as well as rename fields currently named this way.

Comment (Player3324) [reply mappings/net/minecraft/client/render/BuiltChunkStorage.mapping:3]: Right, I didn't even realize we added this to the conventions and I agree this should be fixed separately if desired.

For further discussion in this area it should be noted that the order isn't fixed with a natural scheme, it depends on subject ordering. There would be e.g. xSize and minX. The other form sizeX refers to a size called "x" which makes little sense and xMin implies that the x axis itself had a minimum, which it doesn't.

File: mappings/net/minecraft/client/render/BufferBuilderStorage.mapping
Comment (liach) [inline mappings/net/minecraft/client/render/BufferBuilderStorage.mapping]: Consider improving the name of this method. I recall you renamed this somewhere else

File: mappings/net/minecraft/util/math/Box.mapping
Comment (liach) [inline mappings/net/minecraft/util/math/Box.mapping:7]: why 1 and 2 instead of 0 and 1?

Comment (Runemoro) [reply mappings/net/minecraft/util/math/Box.mapping:7]: Why 0 and 1 rather than 1 and 2? There doesn't seem to be a consensus on how these should be named, there are currently many things named `x2`. I opened an issue here: https://github.com/FabricMC/yarn/issues/941

**Review by modmuss50 — APPROVED** [review]
Im happy with this, im by no means an expert with rendering so getting more reviews in might not be a bad idea.

**Pull request merged**
